File: CS50X/week6/sentimental-credit/credit.py
# TODO
from cs50 import get_int
from sys import exit

number = get_int("Number: ")
text = str(number)
length = len(text)
luhnsum = 0

# Luhn’s Algorithm
## ref from Stackoverflow (https://stackoverflow.com/questions/51121911/iterate-over-every-nth-element-in-string-in-loop-python)
## ref from codeacademy (https://discuss.codecademy.com/t/how-can-i-loop-through-text-starting-from-the-last-character/339346)

## Multiply every other digit by 2, starting with the number’s second-to-last digit, and then add those products’ digits together.
for c in text[-2::-2]:
    digits = int(c) * 2
    for d in str(digits):
        luhnsum += int(d)

## Add the sum to the sum of the digits that weren’t multiplied by 2.
for c in text[::-2]:
    luhnsum += int(c)

## If the total’s last digit is 0 (or, put more formally, if the total modulo 10 is congruent to 0), the number is valid!
if luhnsum % 10 != 0:
    print("INVALID")
    exit(1)

# American Express uses 15-digit numbers, start with 34 or 37
if length == 15 and text[:2] in ["34", "37"]:
    print("AMEX")
    exit(0)

# MasterCard uses 16-digit numbers, start with 51, 52, 53, 54, or 55
elif length == 16 and text[:2] in ["51", "52", "53", "54", "55"]:
    print("MASTERCARD")
    exit(0)

# Visa uses 13- and 16-digit numbers, start with 4
elif (length == 13 or length == 16) and text[0] == "4":
    print("VISA")
    exit(0)

else:
    print("INVALID")
    exit(1)

File: CS50X/week1/pset1/credit/credit.c
#include <cs50.h>
#include <stdio.h>

bool checksum(long card);
long power(int base, int exp);

int main(void)
{
    long card = get_long("Number: ");
    if (card < power(10, 12) || card >= power(10, 16))
    {
        printf("INVALID\n");
    }
    else
    {
        if (checksum(card))
        {
            // printf("Checksum TRUE\n");
            // if 15 digit and start w 34 or 37 then AMEX
            if ((card >= 34 * power(10, 13) && card < 35 * power(10, 13)) ||
                (card >= 37 * power(10, 13) && card < 38 * power(10, 13)))
            {
                printf("AMEX\n");
            }

            // else if 16 digit and start w 51, 52, 53, 54 or 55 then MASTERCARD
            else if (card >= 51 * power(10, 14) && card < 56 * power(10, 14))
            {
                printf("MASTERCARD\n");
            }

            // else if 13 or 16 digit and start with 4 then VISA
            else if ((card >= 4 * power(10, 12) && card < 5 * power(10, 12)) ||
                     (card >= 4 * power(10, 15) && card < 5 * power(10, 15)))
            {
                printf("VISA\n");
            }

            else
            {
                printf("INVALID\n");
            }
        }
        else
        {
            printf("INVALID\n");
        }
    }
}

bool checksum(long card)
{
    int total = 0;
    for (int i = 0; i < 16; i++)
    {
        long place = power(10, i + 1);
        long placeb4 = power(10, i);
        int digit = (card % place - card % placeb4) / placeb4;
        if (i % 2 != 0)
        {
            digit *= 2;
        }
        // printf("%i\n", digit);
        if (digit >= 10)
        {
            int subtotal = 0;
            for (int j = 0; j < 2; j++)
            {
                place = power(10, j + 1);
                placeb4 = power(10, j);
                int subdigit = (digit % place - digit % placeb4) / placeb4;
                subtotal += subdigit;
            }
            digit = subtotal;
        }
        total += digit;
        // printf("TOTAL %i\n", total);
    }
    if (total % 10 == 0)
    {
        return true;
    }
    else
    {
        return false;
    }
}

long power(int base, int exp)
{
    long result = 1;
    while (exp != 0)
    {
        result *= base;
        --exp;
    }
    return result;
}